[FIXED JENKINS-46065] Add extension point for contributing validation

abayer opened the pull request:
* JENKINS issue(s):
    * [JENKINS-46065](https://issues.jenkins-ci.org/browse/JENKINS-46065)
* Description:
    * Adds an extension point for contributing additional validation checks.
* Documentation changes:
    * n/a
* Users/aliases to notify:
    * @reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

rsandell commented:
Possibly getting the run as a parameter would be good to have, to be able to do more validation in particular places, or only do validation on multibranch etc.

rsandell commented:
For example in the case of the `branch` when condition we know that only works for multi branch and when the Jenkinsfile is grabbed from an scm and so we would like to have that info when validating if `branch` can be used or not.

abayer commented:
@rsandell There are validation contexts where we don't have a `run` - i.e., CLI linting or the Editor - so I'd prefer not to have validation that only fires when we're in the actual run. I want consistency there if at all possible.

rsandell commented:
So it is better to fail on stage 45 because the `BRANCH_NAME` env var is missing than failing the build straight away but missing that check when linting over cli?

abayer commented:
@rsandell Lemme think on it...

abayer commented:
I've added `FlowExecution`s as `CheckForNull` parameters to the `DeclarativeValidatorContributor#validateElement` methods. I've also rewritten the logic for [JENKINS-45955](https://issues.jenkins-ci.org/browse/JENKINS-45955) to replace #176 by implementing blocked steps/methods as a `DeclarativeValidatorContributor`. There's a little bit of a harmless loophole in that if a different extension blocks a new method call (i.e., a wrapper step), it'll get past one validation check (`ModelValidatorImpl#validateElement(ModelASTOption)` checks to see if the name is in `Options#getAllowedOptionTypes()`'s keys - and those keys have steps blocked in `BlockedStepsAndMethodCalls#blockedInMethodCalls()` filtered out) but will still fail when `DeclarativeValidatorContributor#validateElement(ModelASTMethodCall, FlowExecution)` gets called for the new extension. Also, the offending step will still show up in the list of possible option names for validation errors. Imperfect, but I couldn't find a more elegant way to pull this off without keeping #176 as well as this, and that seemed silly.

rsandell commented on pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/DeclarativeValidatorContributor.java:
```diff
@@ -65,122 +66,122 @@
      */
     @CheckForNull
-    public final String validateElement(@Nonnull ModelASTElement element) {
+    public final String validateElement(@Nonnull ModelASTElement element, @CheckForNull FlowExecution execution) {
```
Documenting the `execution` would be usefull, esp. when it's `null` and not and why that might matter.
Should be enough to do on a class level so you don't need to cut & paste to all the methods :)

jglick commented on pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy:
```diff
@@ -705,6 +702,26 @@ class ModelValidatorImpl implements ModelValidator {
+
+    private boolean validateFromContributors(ModelASTElement element, boolean isValid, boolean isNested = false) {
+        boolean contributorsValid = DeclarativeValidatorContributor.all().every { contributor ->
+            String error = null
```
unused initializer

jglick commented on pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/validator/ModelValidatorImpl.groovy:
```diff
@@ -705,6 +702,26 @@ class ModelValidatorImpl implements ModelValidator {
+        }
+        if (isValid) {
+            return contributorsValid
+        }
```
`else` what do you `return`?

abayer replied:
ha, that's a mistake.

jglick commented on pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/BlockedStepsAndMethodCalls.java:
```diff
@@ -0,0 +1,87 @@
+        map.put("stage", Messages.BlockedStepsAndMethodCalls_BlockedSteps_Stage());
+        map.put("properties", Messages.BlockedStepsAndMethodCalls_BlockedSteps_Properties());
+        map.put("parallel", Messages.BlockedStepsAndMethodCalls_BlockedSteps_Parallel());
+        return map;
```
`ImmutableMap.of`

jglick commented on pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/BlockedStepsAndMethodCalls.java:
```diff
@@ -0,0 +1,87 @@
+    public static Map<String, String> blockedInMethodCalls() {
+        Map<String, String> map = new LinkedHashMap<>();
+        map.put("node", Messages.BlockedStepsAndMethodCalls_BlockedSteps_Node());
+        map.putAll(blockedInSteps());
```
`Collections.singletonMap`

jglick commented on pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/DeclarativeValidatorContributor.java:
```diff
@@ -0,0 +1,197 @@
+     * Fallback for any unknown element type. Always returns true, cannot be overridden.
+     */
+    @CheckForNull
+    public final String validateElement(@Nonnull ModelASTElement element, @CheckForNull FlowExecution execution) {
```
Using method overloading this way is a very bad idea. Pick a distinct name for each method here.

abayer replied:
Going with method overloading to match up with how `ModelValidator` works.

jglick commented on pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/DeclarativeValidatorContributor.java:
```diff
@@ -0,0 +1,197 @@
+ */
+public abstract class DeclarativeValidatorContributor implements ExtensionPoint {
+    /**
+     * Fallback for any unknown element type. Always returns true, cannot be overridden.
```
So why is it in the API at all?

abayer replied:
Safety - we need to be able to handle cases like a new `ModelASTElement` being added to the validator without breaking compatibility with an extension of `DeclarativeValidatorContributor` that predates that new `ModelASTElement`.

jglick commented on pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/validator/DeclarativeValidatorContributor.java:
```diff
@@ -0,0 +1,197 @@
+ * Extension point for contributing additional {@link ModelASTElement} validation checks to {@link ModelValidator} runs.
+ *
+ * Override a {@code validateElement} method in your extension to have that check run at the end of normal 
+ * Each method should return a string containing the error message if validation fails, and null otherwise.
```
Document the `execution` parameters also.

abayer commented:
@jglick - No downstream uses at this moment, just internal ones.

abayer commented:
Bah, types.
